src/Engines/InputParser.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Jupyter.Core
+{
+    internal class InputParser
+    {
+        public ISymbolResolver Resolver;
+        
+        public InputParser(ISymbolResolver resolver)
+        {
+            this.Resolver = resolver;
+        }
+
+        public bool IsMagicOrHelp(string input, out ISymbol symbol, out string commandInput, out bool isHelp, out string remainingInput)
```

Comment:
In what cases can `input` be `null`? With `#nullable enable` set, it may be worth keeping `input` as non-nullable, then throwing a `NullArgumentException` in the body.